### PR TITLE
tf-head: Fix undefined symbols by linking libtensorflow_cc.so.2

### LINKED ISSUE #3754: With "tf-nightly-cpu==2.12.0.dev20221019" and newer, "import horovod.tensorflow as hvd" throws "undefined symbol" error.
**Environment:**
1. Framework: `TensorFlow`
2. Framework version: `tf-nightly==2.12.0.dev20221019` or newer
3. Horovod version: `All`
4. MPI version: `4.0.3`
5. CUDA version: `N/A`
6. NCCL version: `N/A`
7. Python version: `3.8.10`
8. Spark / PySpark version: `N/A`
9. Ray version: `N/A`
10. OS and version: `Ubuntu 20.04.5 LTS`
11. GCC version: `9.4.0`
12. CMake version: `3.16.3`

**Checklist:**
1. Did you search issues to find if somebody asked this question before? `Yes`
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)? `N/A`
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)? `N/A`
4. Did you check if you question is answered in the [troubleshooting guide] (https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)? `Yes`

**Bug report:**
With "tf-nightly-cpu==2.12.0.dev20221019" and newer, "import horovod.tensorflow as hvd" throws "undefined symbol" as below regardless which version Horovod release is used:

```
(tf-nightly-cpu) root@05635c676813:~# export HOROVOD_VERSION=v0.26.0
(tf-nightly-cpu) root@05635c676813:~# python3 -m pip install --no-cache-dir git+https://github.com/horovod/horovod.git@${HOROVOD_VERSION}
Collecting git+https://github.com/horovod/horovod.git@v0.26.0
  Cloning https://github.com/horovod/horovod.git (to revision v0.26.0) to /tmp/pip-req-build-op6r8nuv
  Running command git clone -q https://github.com/horovod/horovod.git /tmp/pip-req-build-op6r8nuv
  Running command git checkout -q c638dcec972750d4a75b229bc208cff9dc76b00a
  Running command git submodule update --init --recursive -q
Collecting cloudpickle
  Downloading cloudpickle-2.2.0-py3-none-any.whl (25 kB)
Requirement already satisfied: packaging in ./tf-nightly-cpu/lib/python3.8/site-packages (from horovod==0.26.0) (21.3)
Collecting psutil
  Downloading psutil-5.9.3-cp38-cp38-manylinux_2_12_x86_64.manylinux2010_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (295 kB)
     |████████████████████████████████| 295 kB 88.3 MB/s 
Collecting pyyaml
  Downloading PyYAML-6.0-cp38-cp38-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (701 kB)
     |████████████████████████████████| 701 kB 131.0 MB/s 
Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in ./tf-nightly-cpu/lib/python3.8/site-packages (from packaging->horovod==0.26.0) (3.0.9)
Building wheels for collected packages: horovod
  Building wheel for horovod (setup.py) ... done
  Created wheel for horovod: filename=horovod-0.26.0-cp38-cp38-linux_x86_64.whl size=19817679 sha256=58bf0042b9c9d7068d69558ce479884c4dc64feef0618534b361fd9ee3572a88
  Stored in directory: /tmp/pip-ephem-wheel-cache-dha_sfk9/wheels/4d/88/85/e2cf70be97879d9f6662a0e0a2b80d9abc7485bf6f4665af0f
Successfully built horovod
Installing collected packages: cloudpickle, psutil, pyyaml, horovod
Successfully installed cloudpickle-2.2.0 horovod-0.26.0 psutil-5.9.3 pyyaml-6.0
(tf-nightly-cpu) root@05635c676813:~# python
Python 3.8.10 (default, Jun 22 2022, 20:18:18) 
[GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import horovod.tensorflow as hvd
2022-10-22 18:55:02.445270: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/root/tf-nightly-cpu/lib/python3.8/site-packages/horovod/tensorflow/__init__.py", line 27, in <module>
    from horovod.tensorflow import elastic
  File "/root/tf-nightly-cpu/lib/python3.8/site-packages/horovod/tensorflow/elastic.py", line 24, in <module>
    from horovod.tensorflow.functions import broadcast_object, broadcast_object_fn, broadcast_variables
  File "/root/tf-nightly-cpu/lib/python3.8/site-packages/horovod/tensorflow/functions.py", line 24, in <module>
    from horovod.tensorflow.mpi_ops import allgather, broadcast, broadcast_
  File "/root/tf-nightly-cpu/lib/python3.8/site-packages/horovod/tensorflow/mpi_ops.py", line 53, in <module>
    raise e
  File "/root/tf-nightly-cpu/lib/python3.8/site-packages/horovod/tensorflow/mpi_ops.py", line 50, in <module>
    MPI_LIB = _load_library('mpi_lib' + get_ext_suffix())
  File "/root/tf-nightly-cpu/lib/python3.8/site-packages/horovod/tensorflow/mpi_ops.py", line 45, in _load_library
    library = load_library.load_op_library(filename)
  File "/root/tf-nightly-cpu/lib/python3.8/site-packages/tensorflow/python/framework/load_library.py", line 54, in load_op_library
    lib_handle = py_tf.TF_LoadLibrary(library_filename)
tensorflow.python.framework.errors_impl.NotFoundError: /root/tf-nightly-cpu/lib/python3.8/site-packages/horovod/tensorflow/mpi_lib.cpython-38-x86_64-linux-gnu.so: undefined symbol: _ZNK3xla14HloComputation23CollectUnreachableRootsEv
```

I even tried with every single commit after `v0.26.0` that has gone into Horovod's `master` branch and the behavior is the same.
However if `tf-nightly-cpu==2.12.0.dev20221018` is used, this is no loner an issue!
`tf-nightly` also shows the same error if nightly versions above are used.

Chances are the changes that went into TensorFlow's `setup.py` here:
https://github.com/tensorflow/tensorflow/commits/master/tensorflow/tools/pip_package/setup.py
and starting from this commit: `6b7be928f9043c7199810109b1f4d0988c3f0a02`

are contributing to the issue, but I'm not 💯 certain.

### PR DESCRIPTION
**Pull request by maxhgerlach**
## Checklist before submitting

- [x] Did you read the [contributor guide](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md)?
- [ ] Did you update the docs?
- [ ] Did you write any tests to validate this change?  
- [ ] Did you update the [CHANGELOG](https://github.com/horovod/horovod/blob/master/CHANGELOG.md), if this change affects users?

## Description

Fixes #3754 

For context see https://github.com/tensorflow/tensorflow/issues/58281

## Review process to land 

1. All tests and other checks must succeed.
2. At least one member of the [technical steering committee](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md) must review and approve.
3. If any member of the technical steering committee requests changes, they must be addressed.

### CODE REVIEW

**Comment by maxhgerlach** [comment]
This fixes the problem for me. Verified locally:

``` shell
$ pip install -U tf-nightly-cpu==2.12.0.dev20221024
$ pip install -U git+https://github.com/horovod/horovod.git@refs/pull/3755/head
$ python -c 'import horovod.tensorflow'
```

**Comment by maxhgerlach** [comment]
New packages `tf_nightly_{cpu,gpu}-2.12.0.dev20221025` have appeared by now and it looks like the `libtensorflow_cc.so.2` has been removed again. I think the change will be intermittent, so I'll add some conditional to only link the file if it exists.

```
c++: error: /usr/local/lib/python3.8/dist-packages/tensorflow/libtensorflow_cc.so.2: No such file or directory
```
https://buildkite.com/horovod/horovod/builds/8583#01840f11-bd37-4c4e-a85f-4ef0e12838d9

**Comment by maxhgerlach** [comment]
(I verified locally that this works with `tf_nightly_cpu-2.12.0.dev20221024`, so chances are it will continue to work transparently once `libtensorflow_cc.so.2` is re-introduced.)

**Comment by EnricoMi** [comment]
I am not sure if we should keep this while head version does not require this.

File: cmake/Modules/FindTensorflow.cmake [hunk 1]
```diff
@@ -25,6 +25,11 @@ if (LEN EQUAL "4")
     if (Tensorflow_VERSION VERSION_GREATER_EQUAL "2.6")
         # XLA implementations and helpers for resource variables are in _pywrap_tensorflow_internal.so
         set(Tensorflow_LIBRARIES "${Tensorflow_LIBRARIES} ${Tensorflow_LIB_PATH}/python/_pywrap_tensorflow_internal.so")
+        set(lib_tensorflow_cc_so_path "${Tensorflow_LIB_PATH}/libtensorflow_cc.so.2")
```
Comment (EnricoMi) [inline cmake/Modules/FindTensorflow.cmake:28]: what does the `.2` stand for?

Comment (maxhgerlach) [reply cmake/Modules/FindTensorflow.cmake:28]: I hope that stands for TensorFlow 2.x, but I have not dived into the build system to confirm. The main library in that path is `libtensorflow_framework.so.2`. We get the latter via `tf.sysconfig.get_link_flags()`, but I couldn't find an API to resolve the new `libtensorflow_cc.so.2`.

**Comment by maxhgerlach** [comment]
> I am not sure if we should keep this while head version does not require this. We do not intent to support head versions older than today (head of head only if you will).

Agreed. We can always merge this after the rollback of that TensorFlow PR is reverted (comments in https://github.com/tensorflow/tensorflow/pull/55941).

**Comment by maxhgerlach** [comment]
Undefined symbol errors returned to master for tf-head: 
```
E   tensorflow.python.framework.errors_impl.NotFoundError: /usr/local/lib/python3.8/dist-packages/horovod/tensorflow/mpi_lib.cpython-38-x86_64-linux-gnu.so: undefined symbol: _ZTIN10tensorflow11XlaOpKernelE
```
https://github.com/horovod/horovod/actions/runs/3650378067/jobs/6171827788

**Comment by maxhgerlach** [comment]
OK, this PR still fixes the unknown symbols errors.

There's another problem with tf-head / Keras though. I'd suggest to merge this PR and address the other problem separately.

```
___________ ERROR collecting test/parallel/test_tensorflow_keras.py ____________
ImportError while importing test module '/horovod/test/parallel/test_tensorflow_keras.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/lib/python3.8/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
test_tensorflow_keras.py:36: in <module>
    from keras.optimizers.optimizer_v2 import optimizer_v2
E   ModuleNotFoundError: No module named 'keras.optimizers.optimizer_v2'
```

```
___________ ERROR collecting test/parallel/test_tensorflow2_keras.py ___________
ImportError while importing test module '/horovod/test/parallel/test_tensorflow2_keras.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/lib/python3.8/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
test_tensorflow2_keras.py:35: in <module>
    from keras.optimizers.optimizer_v2 import optimizer_v2
E   ModuleNotFoundError: No module named 'keras.optimizers.optimizer_v2'
```

**Pull request merged**